### DISCUSSION TITLE
metadata.metadataBase is not set for resolving social open graph or twitter images

### DISCUSSION THREAD

**Discussion by TommeZ** [vercel/next.js#57251] [Help] [answered]
### Summary

Anyone else got this warning message?

` ⚠ metadata.metadataBase is not set for resolving social open graph or twitter images, using "http://localhost:3000". See https://nextjs.org/docs/app/api-reference/functions/generate-metadata#metadatabase
`

**Comment by eMartin94** [comment, marked as answer]
Add metadataBase:

```
export const metadata = {
  metadataBase: new URL('http://localhost:3000'),
  title: 'Title webtsite',
  description: 'this is the desciption'
  openGraph: {
    title: 'Title webtsite',
    description: 'this is the desciption',
    image: 'url/image.png'
  },
  twitter: {
    card: 'summary_large_image',
    site: '@eMartiiin94',
    title: 'Title webtsite',
    description: 'this is the desciption',
    image: 'url/image.png'
  }
}
```

**Reply by TommeZ** [reply]
This worked. Thanks

**Reply by dev-rafiu** [reply]
This solution worked for me, thanks man!

**Reply by denys-diedov** [reply]
Also is working for me, thank you

**Reply by artginzburg** [reply]
Are you guys suppressing the warning without actually fixing it, or does Next.js automatically replace the explicitly set `new URL('http://localhost:3000')` with the actual URL of the production deployment?

**Reply by eMartin94** [reply]
The error `metadata.metadataBase is not set for resolving` typically occurs when the base URL for metadata is not set, which is needed for resolving social open graph or Twitter images. This warning can appear when running in development mode (localhost), and it will fallback to "http://localhost:3000" if `metadataBase` is not set.

**Reply by jayberry14** [reply]
Where are we meant to add this object in at? Is there a specific file location which it's meant to live at?

**Reply by AlvaroMun** [reply]
Hi @jayberry14, metadata object can be found in layout.js file. Hope this helps you.

**Reply by elhaddajiOtmane** [reply]
where should i add this pls.

**Reply by artginzburg** [reply]
Should this be opened as an issue? I like the automatic behavior that just sets the metadataBase to the deployment origin, and I just would like to not receive a warning about that.

**Reply by jonnylangefeld** [reply]
I'm not surprised the warning goes away on localhost if the value is hard coded to `http://localhost:3000`. But it feels wrong to check this in and deploy to production. Also sometimes, if port 3000 is already used, the app might start on 3001 locally and then it's false again.
After all, `metadataBase` [has a default value](https://nextjs.org/docs/app/api-reference/functions/generate-metadata#default-value), even on localhost. So I agree with @artginzburg, there should be no warning in the first place.

**Reply by artginzburg** [reply]
Sorry if I mislead someone, my bad. Just looked at #49180 to realize that the [default value](https://nextjs.org/docs/app/api-reference/functions/generate-metadata#default-value) is not being set as I expected in production since it uses `VERCEL_URL`, which is an auto-generated URL, not the custom domain assigned to the production deployment.

Went to my website to double-check, and yes, the social image host does not look right (I planned for it to be `rightshift.dev`):
```html
<meta property="og:image" content="https://right-shift-pm2c6zqyo-artginzburg.vercel.app/opengraph-image.png?121e2b7355255853">
```

So, while using `new URL('http://localhost:3000')` is a bad idea, not providing any `metadataBase` is also not a good idea for production as of Next.js 14.0.1

Instead, we can use custom environment variables or just hard-code the production value, which would be OK because apparently `metadataBase` only takes effect in production (see [`resolve-url.ts`](https://github.com/vercel/next.js/blob/7966eaf0f210996abb34d2f02551b3ce28bdfac5/packages/next/src/lib/metadata/resolvers/resolve-url.ts#L14)). We cannot use `headers()` to extract the current URL since they do not exist on build.

**Reply by LudoLogical** [reply]
Given what @artginzburg has said, I want to second the idea that this should be opened as an issue. Expecting everyone using an officially [recommended](https://nextjs.org/docs/app/api-reference/functions/generate-metadata#opengraph) approach for including open graph image metadata to either hard-code a production environment variable or see a warning they can't dismiss doesn't seem reasonable to me.

**Reply by artginzburg** [reply]
@LudoLogical so what should be the issue exactly? That a better default behavior would be to use the current URL origin as the `metadataBase` instead of `VERCEL_URL`, and display no warning in Dev and Preview environments, so that we don't need to set up another environment variable for every new production website?

**Reply by LudoLogical** [reply]
@artginzburg I just sent a response, but I've removed it in favor of starting fresh (sorry for any confusion 😅) because I've just realized what you meant by
> metadataBase only takes effect in production (see [resolve-url.ts](https://github.com/vercel/next.js/blob/7966eaf0f210996abb34d2f02551b3ce28bdfac5/packages/next/src/lib/metadata/resolvers/resolve-url.ts#L14))

One concern I had was that devs should be able to use the following pattern when adding metadata with image URLs:
1. Create the `metadata` export
2. Open `localhost:3000`
3. Find the `<meta>` element(s) in `<head>` and copy the URL(s) from there (notably, these should include `localhost:3000` in development)
4. Paste the `localhost:3000/...` URL(s) in the browser and confirm that the image(s) are as you expect
5. Deploy for the first time (note that if `metadataBase` were applied in dev environments, you would have to do this before steps 3 and 4 since `localhost:3000/...` would instead be `mysite.com/...`)

As you correctly pointed out, `metadataBase` isn't applied in dev environments, so there is no need to do (5) before (3) and (4) and my desired approach works fine. 

That said, as I see it, there are two outstanding problems:
1. The [docs](https://nextjs.org/docs/app/api-reference/functions/generate-metadata#metadatabase) do not make it at all clear that `metadataBase` is ignored outside of production. I would like to propose that we open an issue to clarify them.
2. We still get those warnings in dev environments. What makes the most sense to me is just disabling them in dev and keeping them in preview and production. I think keeping the existing `VERCEL_URL` and current URL origin fallbacks for when `metadataBase` is unspecified is good, but it should probably still cause a warning in production (and maybe in preview too) when you do that.

Assuming both of my suggestions are implemented, the workflow would be very clear:
- Testing metadata URL changes is always possible before deployment in dev environments.
- Production and preview deployments would complain if the fallbacks have to be used. 
  - This might cause a little frustration because "it worked fine when I tested it in dev," but clear docs indicating that `metadataBase` is ignored in dev and definitely not ignored in preview/production should ensure that no one is caught by surprise in the majority of cases. Also, even if it were to cause frustration, the fix would be clear from the docs and would only be encountered once per project (once the `metadataBase` is specified, it's specified, so you'd have to deliberately remove it to get the warnings again).
- Vercel users, users of other managed Next.js deployment services, and self-hosters would all continue to handle any metadata warnings the same way. The existence of Vercel's special `VERCEL_URL` fallback wouldn't preclude non-Vercel users from simply using the current URL origin fallback.

Thoughts? I must admit that, although I'm somewhat familiar with all of this, I'm definitely not a subject matter expert, so there might be a user persona or an edge case I'm not accounting for.

**Reply by jsefiani** [reply]
Did anyone create the issue?

**Comment by SaiRev0** [comment]
![image](https://github.com/vercel/next.js/assets/96655163/458a1506-a3a7-4ece-8a55-21b2a788dbe8)

`⚠ metadata.metadataBase is not set for resolving social open graph or twitter images, using "https://aaspaas-65teeuj9v-sairev0.vercel.app". See https://nextjs.org/docs/app/api-reference/functions/generate-metadata#metadatabase`

I have the same issue in dev mode and in production as well. I don't have any metadataBase variable for the URL. So what should I do ?

**Reply by jpainam** [reply]
@SaiRev0 did you fix this problem?

**Reply by SaiRev0** [reply]
I tried different things but didn't worked for me so at last I added the production URL there, that fixed it `metadataBase: new URL('https://abc.xyz')`

**Reply by Aj7Ay** [reply]
where we have to update this? @SaiRev0

**Reply by SaiRev0** [reply]
@Aj7Ay In your `layout.tsx` file

**Comment by theigor** [comment]
Likewise, I am seeing this in production. A deployment from a few months ago had the metadataBase set correctly and a new prod deployment uses the generated vercel url and is ignoring the base url.

**Comment by jpainam** [comment]
Facing the same issue. I have used
`VERCEL_URL=https://my_url.vercel.app` and 
```typescript
export const metadata = {
  ...metaObject('Appointment'),
};

export const metaObject = (
  title: string,
  openGraph?: OpenGraph,
  description: string = "A default description"
): Metadata => {
  return {
    title: title ,
    description: description,
    metadataBase: new URL('https://my_url.vercel.app'),
    openGraph: openGraph ?? {
      title: title ,
      description: description,
      url: 'https://my_url.vercel.app',
      siteName: '', // https://developers.google.com/search/docs/appearance/site-names
      images: {
        url: 'https://a_valid_amazons3_url.png',
        width: 1200,
        height: 630,
      },
      locale: 'en_US',
      type: 'website',
    },
  };
};
```

But i still get the error when visiting the page.

**Reply by dvzrd** [reply]
Am I to understand that we need to set different url variables for each environment?

>When overriding the default, we recommend using environment variables to compute the URL. This allows configuring a URL for local development, staging, and production environments.
Ref: https://nextjs.org/docs/app/api-reference/functions/generate-metadata#default-value

What if the app is deployed on Vercel, does that mean the defaults are set automatically? Does this include custom domains as well? Sorry, but the docs aren't entirely clear on this. Here's what I ended up doing on one project, but I would rather avoid the extra config files if I can.

In next.config I added basePath env like this:

```js
const basePath = process.env.BASE_PATH ?? '';

module.exports = {
  ...rest of your next config,
  env: {
    basePath,
  },
}
```

Then I setup a metadata config with a url value like this:

```js
export const appMetadata = { 
   ...rest of your metadata config
   url: process.env.basePath || 'https://domain.com',
}
```

Then in app/layout I set the metadataBase along with other metadata defaults for the app:

```js
 metadataBase: new URL(appMetadata.url),
```

There's probably a more optimal way of doing this, I'm sure. Maybe using `@t3-oss/env-nextjs` and `zod` instead of messing with the next config. Seems like there's a better solution to this, would be nice to have an official example.

Does anybody have another solution for setting metadataBase with environment variables that they wouldn't mind sharing?
